strapi/strapi
# [DEITS] add transfer authorization

*** **Draft until we settle on how we want to implement auth, then probably closing this** ***

### What does it do?

- Adds the client side authorization for the `strapi transfer --to` option. The CLI will resolve admin email/password to a login token that is passed to the remote provider (unless the undocumented `toToken` is given in the CLI, in which case those are used)
- adds the server side authorization, authentication (including re-checks for revoked permissions), and permissions
- Adds path and protocol check/conversion
- Lays some of the groundwork for `strapi transfer --from` option

### Why is it needed?

To secure the remote provider websocket using RBAC and an admin login

### How to test it?

When `strapi transfer --to <a strapi admin url>` is run, you should be prompted for email/password which should then log in and use that token for the proceeding transfer.

### TODO:

- [x] check if account has correct permissions
- [x] use `/transfer` route from architecture refactoring branch
- [x] handle URL protocols from CLI (http/https -> ws/wss)
- [x] add permissions in admin
- [x] allow login without prompting (undocumented until approved)
- [ ] tests - lots and lots of tests

Not included in this PR but will come later:

- warning about insecure protocols
- SSO auth -- too complex at this time, but needs to be done at some point
- modify permissions check to transfer.push OR transfer.pull -- first we get push working, then figure that out later
